### ISSUE TITLE
cleanout?

### ISSUE THREAD

**Issue by ddalthorp** [ddalthorp/GenEst#142]
currently the model cannot do other than ignore carcasses found on the 1st day of the monitoring period because those carcasses arrived prior to the period of inference. no need for `removeCleanout` option(?)

**Comment by DAPPERstats** [comment]
if a carcass observation data set with cleanout search carcasses still included is passed to `rghat` or `DWPbyCarcass` it will cause the code to break, so those carcasses need to be removed if they're there. should i just remove the argument option and basically force `removeCleanout` to be `TRUE`?

**Comment by ddalthorp** [comment]
Short answer: Yes.

Longer answer: Currently, the model cannot abide carcasses discovered on the very first search of the monitoring season. But, then, two more questions arise. (1) What about carcasses that arrive before the very first search, are not removed in the "cleanout", and are discovered in a later search? In most cases, this would be a very small number, and we ignore the possibility by assuming that all carcasses discovered after the cleanout arrived during the monitored period. For this simple stab at a solution, we just force removeCleanout to be TRUE. (2) Units that are not searched on the very first monitored day 
 at the site would, on their first search or two, contain a mix of carcasses that arrived before and after the beginning of the monitoring period. To me, it seems that it would be most defensible to throw out some proportion of those carcasses, depending on the persistence distribution and the length of time since the start of the monitoring. However, this adds some complexities that we might best put on the back burner for now because dealing with it in a more robust manner would likely have very little impact on the final estimates.

A related question is what to do about the staggered end of a monitoring season, wherein the final searches occur at different times at different turbines. One idea is would be to ignore carcasses found on the very first search day of the site's monitoring season but include all others and enforce a pseudo-start at t = 0 for all turbines. This would include a few extra carcasses at the start of the year, but they would be counterbalanced by the exclusion of a few carcasses at the end of the year when the final searches at some turbines occurs before the final search of the year. This would likely underestimate M slightly but it should be close. 

Another alternative would be to throw out all carcasses found on each unit's first search and then start the period of inference at the earliest time that all units have been searched. Aside from being more cumbersome to work with, it is bound to be less stable (estimate will depend on that precise date of the first search at the last-searched turbine), less accurate (losing the balance between counting too many at the start and too few at the end), and less satisfying in that we end up throwing out more data. 

So, a (possibly temporary) solution would be to ignore all carcasses found at t = 0 and append t = 0 to  search schedules that begin later. The effect would be to make an implicit assumption that all carcasses found after t = 0 arrived after t = 0. This assumption could be enforced in the field by tossing out carcasses that are determined to have arrived prior to t = 0 (doing a DS729-type thing on a very limited set of carcasses). Or the slight number of extra carcasses at the beginning can be rationalized as counterbalancing the slightly artificially low number of carcasses at the end of the season. This solution would be easily amended to count initial searches as more precise discounted cleanouts...throwing out _some_ of the carcasses found in the first search at a unit, depending on when that search was conducted (i.e., how long after t = 0), and also account for the possibility that some fraction of carcasses found after the first search at a unit arrived earlier than t = 0. This is doable, but it would need to be countered by a similar scheme at the end of the monitoring season to account for units whose searches end earlier than others.

**Comment by ddalthorp** [comment]
Currently, the model crashes if there is a "staggered" cleanout (wherein the first search at any unit is considered a 'cleanout') because it erases the notion of a clean period of inference, which, as formulated, our model requires. 

Immediate kludgy solution: 
1. "cleanout" carcasses on searches at t = 0 for the site as a whole but do not "clean out" carcasses from units that are not first searched until a time later than t = 0.

2. set the t = 0 row of the search schedule to all 1s, which defines t = 0 as a common starting point for inference period at all units.

Effect of the kludge:
Incorrectly assigning a small number of carcasses found at the beginning of the monitoring season to the period of inference beginning at t = 0. This overcount at the beginning of the season is (partially) counterbalanced by an under count at some units near the end of the season. Overall, this will introduce very little error (negligible?) into the estimates.

Longer-term solution:
Better accounting for "gaps" at the beginning and end of the monitored period. The idea is to define t = 0 as the beginning of the inference period and apply a progressive discounting of carcasses found early in the year (so carcasses found at t = 0 are entirely tossed out, carcasses found on a 1st search at a turbine at t > 0 are "mostly" tossed out, etc.). And handle the gaps at the end of the year via dwp-like extrapolation in time (DWT?) rather than space (DWA?). This will take a few days of work to hammer out the details, but it's a minor issue that we should not allow ourselves to be held us up by and go with the kludge for now.

**Comment by DAPPERstats** [comment]
ok, so it sounds like you've got a bead on how to address this from a math perspective, are you able to manage the coding edits to accomplish it?

**Comment by ddalthorp** [comment]
The kludge is easy...and probably easier for you than for me because you are much more in touch with the current shapes and names of the data structures. 

The longer-term solution we will not likely be able to address this week.

**Comment by DAPPERstats** [comment]
where does this stand after your most recent edits? is the need still the same?

**Comment by ddalthorp** [comment]
Yes. The issue remains.

We need a well-defined period of inference, which the kludgy solution outlined above does simply and cleanly (immediate priority). It works very well in "normal" situations but something fancier may be needed to handle the odd, marginal cases (not an immediate priority).

**Comment by DAPPERstats** [comment]
ok, so just to make sure I understand what to do...

`whichCleanout` notes which carcasses are from cleanout searches. cleanout searches are currently defined as the first search in a unit. cleanout searches should be defined as those occurring on t = 0.

`dateToDay` is what's used to set up the day vectors. it's currently being fed the global initial day as t = 0, but it should be using the unit-specific initial day as t = 0, so everything is searched on t = 0, even if those days are offset from each other on the calendar.

does that encompass when needs to be done?

**Comment by ddalthorp** [comment]
no.

Define the start of the period of inference as the global t = 0, i.e., the day at which the very first turbine at the site is searched. To "enforce" that assumption, "clean out" the carcasses found at all turbines searched on that day. For other turbines, set searches_unit = 1 for that global t = 0.

E.g., suppose in CO you have:
```
Unit  date   size
1     apr01   sml
1     apr01   bat
2     apr04   lrg
1     apr08   bat
2     apr18   sml
```
and in SS parameters you have:
```
date0 = apr01
days = c(0, 3, 7, 10, 14, 17, ...)
searches_unit[1, ] = c(1, 0, 1, 0, 1, ...)
searches_unit[2, ] = c(0, 1, 0, 1, 0, ...)
```
Then, toss out only the apr01 carcasses and set searches_unit[2, 1] = 1. This makes the implicit assumption that all carcasses discovered after t = 0 arrived after t = 0. This way, we are slightly over-counting carcasses at the beginning of the year and slightly under-counting carcasses at the end of the year. `dateToDay` would be calculated with apr01 as the reference date for all units.

This is by far the least dangerous of the post-g extrapolations, and, in almost any normal situation, the expected error is very small in comparison to the other post-g adjustments (viz., dwp and sampling fraction).

**Comment by DAPPERstats** [comment]
ah, so rather than shift the search schedule of the unit with a delayed first search back in time, we're just adding a search to that unit on the first possible day.

after those faux-searches are added, cleanout searches end up being **both** the first search at the unit and the search that's at t = 0 in the unit, but if a unit wasn't actually searched on t = 0, then nothing gets removed because nothing was actually found on that search because it was a faux-search.

do i have it finally?

and this needs to be done in `rghat` and `DWPbyCarcass`, right? anywhere else?

**Comment by ddalthorp** [comment]
That sounds about right

_Where_ this needs to be done is an interesting question. In addition to rghat, the splits code expects array sizes to be ncarc x nsim. One issue we were having earlier was that there were arrays of different dimensions floating around depending on what type of cleanout had been performed. For temporal splits, we need SS, and hopefully it would be pre-formatted with searches_unit[i, 1] = 1. We could probably do that in the `SS` object builder function, but then `calcRate` would need to adapt to the `days`/`searches` option. Then...more natural might be to do it at the start of `calcRate`.

**Comment by DAPPERstats** [comment]
great!

yeah, so i've avoided manipulating the data sets outside of the main functions, so that we can pass the user's data directly into functions as quickly as possible (at the command line and within the app), and so that we can insure that all functions are starting from the same data tables. this does assume that each main function can manage its own data management and that they're all working similarly (hence the need to make edits to `rghat` and `DWPbyCarcass`)

**Comment by ddalthorp** [comment]
OK. So the splits functions do not perform cleanouts. But they should, right? This would involve subsetting arrays as necessary to make sure all align right with ghat and Aj. (or some other modification)

**Comment by DAPPERstats** [comment]
yes they should. 
i've got the update integrated for `rghat` and `DWPbyCarcass` and am waiting on the travis checks to clear (fingers crossed), but will merge that in shortly 
i'm planning to work through the splits code now that that's all ready to go finally and will be able to add in the removal of the cleanouts

basically it involves updating the `data_SS` object (by switching dates to days and then adding a `1` into any units where there was a `0` on day 0) and the `data_CO` object (by removing anything found on `t = 0`)

**Comment by DAPPERstats** [comment]
the cleanout situation has been addressed in `calcSplits` v 0.0.5.17 with the PR #187

I had to make some minor changes to the arguments to `calcSplits` (see #187), but I've also updated the vignettes, so you can see how it's implemented now 

I'm going to close this issue now that we've got the short-term solution implemented. The longer term solution can get its own issue
